> _A Man is like a fraction whose numerator is what he is and whose denominator is what he thinks of himself. The larger the denominator, the smaller the fraction._  
> **Tolstoy **

## Exercice 1 : Factorisation
* On intègre ici un test de divisibilité directement par les petits nombres premiers, ce qui évite de tester tous les entiers de $2$ à $257$, mais seulement les premiers. Ces petits nombres premiers sont dans une liste.
* On intègre aussi un hypothétique test de primalité rapide pour envisager où l'utiliser.

In [3]:
def isprime(n):
    return False
    # imaginons que l'on dispose d'une vraie fonction
    # isprime, très très rapide !!!

   
def factor(n):
    """ Retourne une liste de couples (p, e),
    n étant égal au produit des p^e
    Exemple : factor(12) retourne
              [(2, 2), (3, 1)]
    Complexité : O(√(n))
    """
    
    smallPrimes = [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101,103,107,109,113,127,131,137,139,149,151,157,163,167,173,179,181,191,193,197,199,211,223,227,229,233,239,241,251,257]
    lastSmallPrime = smallPrimes[-1]
    
    assert n>0
    Factor = []
    
    # on écrème les petits facteurs de n
    for p in smallPrimes:
        if n%p == 0:
            n //= p
            e = 1
            while n%p == 0:
                n //= p
                e += 1
            Factor.append((p, e))
            if n == 1:
                return Factor
            if p*p > n:
                # ici n est premier
                Factor.append((n, 1))
                return Factor
            
    # arrivé là, n ne possède plus de petits facteurs premiers
    # on fait alors un test de primalité rapide
    
    if isprime(n):
        Factor.append((n, 1))
        return Factor
    
    # arrivé là, n est composé,
    # avec des facteurs 'moyens' ou 'grands'
    
    p = lastSmallPrime
    while n>1:
        p += 2
        if p*p > n:
            # dans ce cas, n est égal à 1, ou bien premier
            if n > 1:
                Factor.append((n, 1))
            return Factor
        if n%p == 0:
            n //= p
            e = 1
            while n%p == 0:
                n //= p
                e += 1
            Factor.append((p, e))
            if isprime(n):
                Factor.append((n, 1))
                return Factor

    return Factor

for n in range(7, 13):
    print(f"factor({n}) retourne :", factor(n))

for n in [4953851, 600851475143, 14837457737, 2000000018]:
    print(f"factor({n}) retourne :", factor(n))


factor(7) retourne : [(7, 1)]
factor(8) retourne : [(2, 3)]
factor(9) retourne : [(3, 2)]
factor(10) retourne : [(2, 1), (5, 1)]
factor(11) retourne : [(11, 1)]
factor(12) retourne : [(2, 2), (3, 1)]
factor(4953851) retourne : [(7, 2), (17, 1), (19, 1), (313, 1)]
factor(600851475143) retourne : [(71, 1), (839, 1), (1471, 1), (6857, 1)]
factor(14837457737) retourne : [(1471, 2), (6857, 1)]
factor(2000000018) retourne : [(2, 1), (1000000009, 1)]


---
## Exercice 2 - SPOJ - [TEST](http://www.spoj.com/problems/TEST/)
_Rewrite small numbers from input to output. Stop processing input after reading in the number 42._ 

In [1]:
while True:
    n = int(input())
    if n == 42: break
    print(n)

1
1
2
2
88
88
42


---
## Exercice 3 - SPOJ - [SMPDIV](http://www.spoj.com/problems/SMPDIV/)
**Divisibility**

_Print all integers $a_i$ such that $a_i$ is divisible by $x$ and not divisible by $y$, where $1 < a_i < n < 100\,000$._

In [3]:
for t in range(int(input())):
    line = input() # lecture de la ligne entière
    n, x, y = line.split() # n, x, y sont au format texte
    n, x, y = int(n), int(x), int(y) # changement de type à entier
    for a in range(0, n, x):
        if a%y != 0: print(a, end=" ")
    print()

2
7 2 4
2 6 
35 5 12
5 10 15 20 25 30 


In [4]:
# avec une écriture plus fonctionelle
def main():
    for t in range(int(input())):
        n, x, y = map(int, input().split())
        print(" ".join(map(str, (a for a in range(0, n, x) if a%y))))
main()

2
7 2 4
2 6
35 5 12
5 10 15 20 25 30


---
## Exercice 3 - SPOJ - [STRHH](http://www.spoj.com/problems/STRHH/)
_Given a sequence of $2k$ characters, please print every second character from the first half of the sequence.
Start printing with the first character._


In [5]:
def main():
    for t in range(int(input())):
        line = input()
        l = len(line)
        print(line[:l//2:2])
main()

4
your
y
progress
po
is
i
noticeable
ntc


---
## Exercice 4 - SPOJ - [ALCATRAZ1](http://www.spoj.com/problems/ALCATRAZ1/)
_You are being given a number $N$ . $(1\leqslant N \leqslant 10^{50})$ . You have to print the sum of digits of that particular number._

In [7]:
def main():
    for t in range(int(input())):
        N = int(input())
        S = 0
        while N!=0:
            N, r = divmod(N, 10)
            S += r
        print(S)
main()

3
123123123
18
3434
14
1234567890
45


**Critique du script :** C'est dommage de lire l'entrée, la transformer en binaire, pour ensuite la transformer à nouveau en décimal...

In [1]:
def main():
    for t in range(int(input())):
        N = input().strip() # strip enlève les espaces en trop qu'il pourrait y avoir
        S = sum(ord(x)-ord('0') for x in N)
        print(S)
main()

3
123123123
18
1234567890
45
5
5


On peut ne faire qu'une seule soustraction, au lieu d'une par boucle.  
On peut appliquer la méthode `map` à la fonction `ord` sur la chaîne `N`.

In [3]:
def main():
    for t in range(int(input())):
        N = input().strip() # strip enlève les espaces en trop qu'il pourrait y avoir
        print( sum(map(ord, N)) - ord('0')*len(N) )
main()

1
1234567890
45


---
## Exercice 5 - SPOJ - [PRIME1](http://www.spoj.com/problems/PRIME1/)
_Your task is to generate all prime numbers between two given numbers!_

On va construire une version modifiée du crible d'Ératosthène : le crible segmenté ; tranche par tranche.
* Pour $1<m\leqslant n$, dans l'intervalle $[\![m ; n]\!]$, les nombres qui n'ont aucun facteur premier entre $2$ et $\sqrt{n}$ sont tous premiers.
* On construit d'abord tous les nombres premiers jusqu'à $\sqrt{n_\text{max}}$. Les _petits_ nombres premiers.
* On utilise ces facteurs pour faire le crible dans $[\![m ; n]\!]$,
* on utilise une liste `Crible2` telle que `Crible2[i] == 1` pour `m+i` premier, `0` sinon.
* Pour chaque _petit_ premier `p`, on cherche le premier multiple `kp` de `p` dans $[\![m ; n]\!]$,
* s'il est différent de `p`, c'est un composé, comme les suivants, qui sont de la forme `kp + c*p`, ...

In [4]:
def main():
    # première étape ; voir D-Exos-corrigés-2
    lim = 31622 # la racine carrée de n_max (10⁹)
    Crible = [1]*lim
    Crible[:2] = [0]*2
    p = 2
    while p*p<lim:
        if Crible[p] == 1:
            # p est alors premier,
            # on 'coche' les multiples de p comme composés,
            # il suffit de commencer à p²
            for k in range(p*p, lim, p):
                Crible[k] = 0
        p += 1
    # On construit alors la liste :
    Prime = []
    for p in range(lim):
        if Crible[p] == 1:
            Prime.append(p)
    #==========================
    
    # deuxième étape
    
    for _ in range(int(input())):
        m, n = map(int, input().split())
        n += 1 # on ira de m inclus, jusqu'à n exclu !
        Crible2 = [1]*(n-m)
        if m==1:
            Crible2[0] = 0
            # 1 n'est pas premier
        for p in Prime:
            if p*p > n:
                break
                # une sortie prématurée
                # dans ce cas, Crible2 est fini
            kp = (m-1) - (m-1)%p
            # kp est un multiple de p, kp < m, le plus grand
            kp += p
            # kp est alors un multiple de p, kp >= m, le plus petit 
            if kp == p:
                kp += p
            # kp est désormais un multiple distinct de p, donc composé
            for c in range(kp, n, p):
                Crible2[c-m] = 0
                # on marque les entiers c comme composé
        
        # On génère les premiers dans [m, n]
        
        for i in range(n-m):
            if Crible2[i] == 1:
                print(m+i)
                
main()

2
1 10
2
3
5
7
3 5
3
5


* Ce code permet d'entrer dans la page 3 du 'ranking' section Python3,
* mais il n'est pas assez rapide pour entrer page 1 !!!
> Pour faire d'autres expériences, considérer [PRINT](http://www.spoj.com/problems/PRINT/), ou le code ci-dessus sera trop lent,  
> il faudra trouver des optimisations. Conseil : penser à `bytearray` et `compress`, ...  
> Attention, il faut une bonne maîtrise de Python pour cela.

---
## Exercice 5 - SPOJ - [AMR11E](http://www.spoj.com/problems/AMR11E/)
Les entiers ayant au moins trois facteurs premiers distincts.
* Commençons par une méthode force brute, mais on se doute que ces entiers deviennent très fréquents
* On en déduira qu'ici, établir la liste des $1000$ premiers tels nombres sera rapide


In [8]:
def factor(n):
    """ Retourne une liste de couples (p, e),
    n étant égal au produit des p^e
    Exemple : factor(12) retourne
              [(2, 2), (3, 1)]
    Complexité : O(√(n))
    """
    
    assert n>0
    Factor = []
    p = 2
    while n>1:
        # début modification
        if p*p > n:
            # dans ce cas, n est égal à 1, ou bien premier
            if n > 1:
                Factor.append((n, 1))
            return Factor
        # fin modification
        if n%p == 0:
            n //= p
            e = 1
            while n%p == 0:
                n //= p
                e += 1
            Factor.append((p, e))
        p += 1
    return Factor

def main():
    Lucky = [0, 30, 42]
    # ajout de 0, pour avoir une liste indexée à partir de 1.
    # Lucky[1] = 30 ; l'élément n°1 est 30
    Q = 2 # on en a déjà 2 donné ; cadeau 
    l = 42 # le point de départ
    while Q < 1000:
        l += 1
        if len(factor(l)) >= 3:
            Lucky.append(l)
            Q += 1
    
    for _ in range(int(input())):
        n = int(input())
        print(Lucky[n])
main()

3
1
30
2
42
1000
2664


On constate qu'en effet, le 1000ᵉ élément est 2664, et que la force brute permet de vite factoriser tous les entiers jusqu'à 2664.  
Un problème plus intéressant pourrait être : Afficher tous les entiers jusqu'à une certaine limite (comme 2 millions) qui n'ont que deux diviseurs premiers.

---
## Exercice 5 - SPOJ - [ENIGMATH](http://www.spoj.com/problems/ENIGMATH/)
_For a given value of $A$ and $B$, find the minimum positive integer value of $x$ and $y$ that satisfies the equation :_
$$Ax - By = 0$$

---
On note $D$ le PGCD de $A$ et $B$, on a $A = DA'$, et $B=DB'$, avec $A'$ et $B'$ entiers premiers entre eux.

* $A$ et $B$ sont non nuls, donc $D\neq0$.
* Les solutions de notre équation sont donc exactement les mêmes que l'équation : $A'x = B'y$
* D'après le lemme de Gauss, sachant que $A' \land B' = 1$ :  
    * $x$ est un multiple de $B'$
    * $y$ est un multiple de $A'$
    
* $x=B'$ et $y=A'$ est une solution acceptable évidente, et donc la plus petite en entiers strictement positifs.
> **Note** : En anglais, _positive_ signifie « strictement positif ».  
> Pour traduire « positif ou nul », on emploie _non negative_.



In [9]:
from math import gcd
def main():
    for _ in range(int(input())):
        a, b = map(int, input().split())
        d = gcd(a, b)
        print(b//d, a//d)
main()

1
2 3
3 2


---
## Project Euler 23 - [Non-abundant sums](https://projecteuler.net/problem=23)
_A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of $28$ would be $1 + 2 + 4 + 7 + 14 = 28$, which means that $28$ is a perfect number._

_A number $n$ is called deficient if the sum of its proper divisors is less than $n$ and it is called abundant if this sum exceeds $n$._

_As $12$ is the smallest abundant number, $1 + 2 + 3 + 4 + 6 = 16$, the smallest number that can be written as the sum of two abundant numbers is $24$. By mathematical analysis, it can be shown that all integers greater than $28123$ can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit._

_Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers._


In [2]:
def sumDiv(n):
    """Somme des diviseurs propres de n"""
    if n<2: return 0
    if n<4: return 1
    S = 1
    for k in range(2, n):
        if k*k>=n:
            if k*k == n:
                S += k
            return S
        l, r = divmod(n, k)
        if r == 0:
            S += k+l       

ans = 0
abondant = set()
for x in range(1, 28124):
    if sumDiv(x) > x:
        abondant.add(x)
        
    #=== variante longue
    sum2a = False
    for y in abondant:
        if x-y in abondant:
            sum2a = True
            break
    if not sum2a:
        ans += x
    #=== variante courte
    #if not any(x-y in abondant for y in abondant):
    #    ans += x
    #=== fin variantes

print(ans)

4179871


---
# Devoirs pour la semaine suivante
1. SPOJ [Sum](http://www.spoj.com/problems/UCBINTA/)
2. SPOJ [Divisor Summation](http://www.spoj.com/problems/DIVSUM/)
3. SPOJ [Prime After N](http://www.spoj.com/problems/AU12/)
4. SPOJ [n-divisors](http://www.spoj.com/problems/NDIV/) Pour les plus motivés.
5. Project Euler [Truncatable primes](https://projecteuler.net/problem=37)

## Exercice 6
**Un problème indispensable** : ici on jette les bases d'une méthode pour factoriser très rapidement une grande quantité de 'petits' nombres.

Construire la liste ```factorMini``` **indexée à zéro**, de longueur un million, telle que :
* ```factorMini[0] = 0```, inutilisé.
* ```factorMini[1] = 1```
* ```factorMini[n] = p```, où $p$ est le plus petit facteur premier de $n>1$.

Cette [liste](https://oeis.org/A020639) commence donc par : ```[0, 1, 2, 3, 2, 5, 2, 7, 2, 3, ...```

Par exemple, en 9<sup>ème</sup> position, $9$ est composé et son plus petit facteur premier est $3$. 

---

**Conseil :** s'inspirer du crible d'Ératosthène. Voici des pistes.

L'objectif ici n'est pas de factoriser un nombre, mais de nombreux. Dans ce cas, l'utilisation d'une liste est très utile.

* On initialise une liste $u_n = n$.
* On itère sur les premiers termes, et si $u_p$ est resté à $p$, c'est qu'il est premier, dans ce cas, chacun de ses multiples est à traiter :
    * ou bien, il avait déjà un facteur premier signalé, on le laisse,
    * ou bien, on indique que $p$ est le plus petit facteur premier de cet entier. 
